drivers: regulator: Remove non-const usage of config struct

danieldegrasse opened the pull request:
Config struct in pmic regulator driver was being used as a non-constant
value when an array defined at compile time was cast to a struct used at
runtime. Move this pointer to the data struct in pmic driver.

Also removes unnecessary NULL check in PMIC initialization function (since `device_is_ready` handles NULL devices correctly)

Fixes #41951

danieldegrasse committed:
drivers: regulator: Remove non-const usage of config struct

Config struct in pmic regulator driver was being used as a non-constant
value when an array defined at compile time was cast to a struct used at
runtime. Move this pointer to the data struct in pmic driver.

Fixes #41951

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

gmarull commented on drivers/regulator/regulator_pmic.c:
```diff
@@ -286,14 +291,15 @@ static int disable_regulator(const struct device *dev)
-	config->current_levels = (struct current_range *)config->current_array;
+	data->current_levels = (struct current_range *)config->current_array;
 	/* Check to verify we have a valid I2C device */
 	if (config->i2c_dev == NULL || !device_is_ready(config->i2c_dev)) {
```
nit: not related, but NULL check here can be removed (device_is_ready supports NULL as an input)

danieldegrasse replied:
Thanks for the heads up, I added a commit to this PR to fix that

gmarull reviewed (approved):
Thanks, LGTM

danieldegrasse committed:
drivers: regulator: pmic: remove unneeded NULL check

Remove unneeded NULL check in pmic init function. Not required as
device_is_ready() will handle NULL device structs.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>
